src/metabase/api/database.clj

```diff
@@ -206,29 +206,25 @@
      filtered-dbs)))
 
 (defn- check-db-data-model-perms
-  [db include-editable-data-model?]
-  (if include-editable-data-model?
-    (let [filtered-dbs (filter-databases-by-data-model-perms [db])]
-      (when (seq filtered-dbs)
-        (first filtered-dbs)))
-    db))
-
-(defn- dbs-list [& {:keys [include-tables?
-                           include-saved-questions-db?
-                           include-saved-questions-tables?
-                           include-editable-data-model?
-                           exclude-uneditable-details?]}]
+  "Given a DB, checks that *current-user* has any data model editing perms for the DB. If yes, returns the DB,
+  with its tables also filtered by data model editing perms. If it does not, throws a permissions exception."
+  [db]
+  (let [filtered-dbs (filter-databases-by-data-model-perms [db])]
+    (api/check-403 (first filtered-dbs))))
+
+(defn- dbs-list
+  [& {:keys [include-tables?
+             include-saved-questions-db?
+             include-saved-questions-tables?
+             include-editable-data-model?
+             exclude-uneditable-details?]}]
   (let [dbs (db/select Database {:order-by [:%lower.name :%lower.engine]})
-        dbs (if include-editable-data-model?
-              ;; Since users can have *data model* perms even if they have no *data* perms for a DB, we don't filter by
-              ;; read permissions here if `include-editable-data-model?` is true. Instead, we only return ID, name
-              ;; and tables for DBs for which the user has no data perms, to avoid exposing unecessary DB metadata.
-              dbs
-              (filter mi/can-read? dbs))
-        dbs (if exclude-uneditable-details? (filter mi/can-write? dbs) dbs)]
+        filter-by-data-access? (not (or include-editable-data-model? exclude-uneditable-details?))]
     (cond-> (add-native-perms-info dbs)
       include-tables?              add-tables
       include-editable-data-model? filter-databases-by-data-model-perms
+      exclude-uneditable-details?  (#(filter mi/can-write? %))
+      filter-by-data-access?       (#(filter mi/can-read? %))
```

Comment:
Cleaned up the semantics here: we _either_ filter the DB list by data access (i.e. `mi/can-read?`) when powering the main app, _or_ by detail editing access (i.e. `mi/can-write?`) when powering the admin panel. Before, `mi/can-read?` checked both data access and details editing permissions, which muddled things up a lot.